### PR TITLE
Add APIs to atomically replace a file

### PR DESCRIPTION
**Pull request by cgwalters**
We have similar code inlined in coreos/bootupd; let's
generalize it here.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -176,6 +230,106 @@ pub(crate) fn impl_ensure_dir_all(d: &openat::Dir, p: &Path, mode: libc::mode_t)
...
+
+    /// Flush any outstanding buffered data and rename the temporary
+    /// file into place.  The provided closure will be invoked
+    /// on the real underlying file descriptor before it is
+    /// renamed into place.  You can use this to change file attributes.
+    /// For example, you can change the mode, extended attributes, or invoke
+    /// `fchmod()` to change ownership, etc.
+    pub fn complete_with<F>(mut self, f: F) -> io::Result<()>
+    where
+        F: Fn(&fs::File) -> io::Result<()>,
```
Comment (lucab) [inline src/lib.rs:310]: `Fn(&mut impl std:io::Write)` perhaps?

**Review by lucab — COMMENTED** [review]
Two high-level things:
 * it looks like this PR cares a lot about finalizing writes, but safe code is allowed to bypass the safety logic in destructors anyway (e.g. with `mem::forget()`). To that extent, only keeping `complete_with()` may be enough.
 * I guess that you don't want consumers to mess with your writer/file mid-flight, thus the internal BufWriter sould be hidden instead, maybe?

**Comment by cgwalters** [comment]
> it looks like this PR cares a lot about finalizing writes, but safe code is allowed to bypass the safety logic in destructors anyway (e.g. with mem::forget()). To that extent, only keeping complete_with() may be enough.

The goal here is to solve the same problem outlined in [the drop_bomb crate](https://docs.rs/drop_bomb/0.1.5/drop_bomb/) - it's too easy to forget the `flush()` and omit error checking.

Leaking the writer, sure, that's a bug in the calling code; nothing we can or should do about it right?  Not the problem I'm trying to solve.

> I guess that you don't want consumers to mess with your writer/file mid-flight, thus the internal BufWriter sould be hidden instead, maybe?

In contrast, the consumer owns the `FileWriter` and its internal writer for the duration until they pass ownership back into `complete()/abandon()`.  

Basically I want callers to be able to directly use the writer.  (Why a `BufWriter`?  Well...maybe it should be configurable but it's a sensible default; as you pointed out once in my code omitting the buffering is another Rust footgun.  If the caller really wants they can [bypass the buffering](https://doc.rust-lang.org/std/io/struct.BufWriter.html#method.get_mut) but I can't think of a reason for that)

File: src/lib.rs (hunk 1 above)
Comment (cgwalters) [reply src/lib.rs:310]: No, per the docs the caller must be able to access a real file descriptor to invoke `fchown/fchmod` etc.

**Comment by cgwalters** [comment]
One important thing to keep in mind is the goal of this crate is to be high level helpers for the "raw" [openat crate](https://docs.rs/openat/0.1.19/openat/) - if someone wants to do something unusual they can use those APIs directly.

This particular PR is about correctly implementing the "atomically replace a file" pattern in a convenient way.

**Comment by cgwalters** [comment]
Now one thing you did touch on here is that we could *probably* get away with just the `write_file_with()` API and omit making `FileWriter` public but...eh, that makes error handling messier.

Unless hm, I guess we could make it generic on the error type too and not just `io::Error`.

**Comment by cgwalters** [comment]
> Unless hm, I guess we could make it generic on the error type too and not just io::Error.

Hmm right but then we have the opposite problem, we have an `io::Error` but need to return the caller's type...trying this but my generics-fu isn't quite good enough

```
diff --git a/src/lib.rs b/src/lib.rs
index 5de3b4c..044a1db 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -65,24 +65,28 @@ pub trait OpenatDirExt {
 
     /// Atomically create or replace the destination file, calling the provided
     /// function to generate the contents.
-    fn write_file_with<P: AsRef<Path>, F>(
+    fn write_file_with<P: AsRef<Path>, F, T, E>(
         &self,
         destname: P,
         mode: libc::mode_t,
         f: F,
-    ) -> io::Result<()>
+    ) -> Result<T, E>
     where
-        F: Fn(&mut std::io::BufWriter<std::fs::File>) -> io::Result<()>,
+        F: Fn(&mut std::io::BufWriter<std::fs::File>) -> Result<T, E>,
+        E: Into<Box<dyn std::error::Error + Send + Sync>>,
     {
         let mut w = self.new_file_writer(destname, mode)?;
         match f(&mut w.writer) {
-            Ok(()) => {}
+            Ok(v) => {
+                w.complete()?;
+                Ok(v)
+            },
             Err(e) => {
                 w.abandon();
-                return Err(e);
+                Err(e)
             }
         }
-        w.complete()
+       
     }
 
     /// Atomically create or replace the destination file with
```

:arrow_down: 

```
walters@toolbox /v/s/w/s/g/c/openat-ext (file-writer)> cargo build
   Compiling openat-ext v0.1.6 (/var/srv/walters/src/github/cgwalters/openat-ext)
error[E0277]: `?` couldn't convert the error to `E`
  --> src/lib.rs:78:57
   |
78 |         let mut w = self.new_file_writer(destname, mode)?;
   |                                                         ^ the trait `std::convert::From<std::io::Error>` is not implemented for `E`
   |
   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
   = note: required by `std::convert::From::from`
help: consider further restricting this bound
   |
76 |         E: Into<Box<dyn std::error::Error + Send + Sync>> + std::convert::From<std::io::Error>,
   |                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: `?` couldn't convert the error to `E`
  --> src/lib.rs:81:29
   |
81 |                 w.complete()?;
   |                             ^ the trait `std::convert::From<std::io::Error>` is not implemented for `E`
   |
   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
   = note: required by `std::convert::From::from`
help: consider further restricting this bound
   |
76 |         E: Into<Box<dyn std::error::Error + Send + Sync>> + std::convert::From<std::io::Error>,
   |                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
error: could not compile `openat-ext`.

To learn more, run the command again with --verbose.
walters@toolbox /v/s/w/s/g/c/openat-ext (file-writer) [101]> 
```

File: src/lib.rs (hunk 1 above)
Comment (lucab) [reply src/lib.rs:310]: My bad, I totally missed the docstring and also misunderstood the intended usage of this.

**Comment by lucab** [comment]
From an out-of-band chat: the current design of this makes sense as-is. It is not meant to prevent people from abusing it in weird ways, but to make the happy path easier to be performed correctly.
Regarding the error type, a `Box<std::error::Error>` could help with consumers' error types, but given the context above not forcing everyone to go through a closure is possibly a better / more ergonomic approach anyway.

**Comment by cgwalters** [comment]
> Regarding the error type, a Box<std::error::Error> could help with consumers' error types, but given the context above not forcing everyone to go through a closure is possibly a better / more ergonomic approach anyway.

Right, though I just figured out the generic sorcery :mage: necessary to have the closure path support generic error types too; basically we needed `E: From<io::Error>` so we can pass back our error types too.

So that's strictly better and will allow most cases to use the closure path which is great.

The `FileWriter` though is still necessary for more control - things like the file prefix/suffix as well as final changes to the file descriptor.

**Comment by cgwalters** [comment]
OK I now added an explicit `with_sync()` variant so the bootupd patch is just:

```
diff --git a/src/bootupd.rs b/src/bootupd.rs
index 40c0fb2..674ebf1 100644
--- a/src/bootupd.rs
+++ b/src/bootupd.rs
@@ -9,7 +9,6 @@ use openat_ext::OpenatDirExt;
 use serde::{Deserialize, Serialize};
 use std::borrow::Cow;
 use std::collections::BTreeMap;
-use std::io::prelude::*;
 use std::path::Path;
 
 /// Stored in /boot to describe our state; think of it like
@@ -187,34 +186,10 @@ pub(crate) fn validate(name: &str) -> Result<ValidationResult> {
 /// Atomically replace the on-disk state with a new version
 fn update_state(sysroot_dir: &openat::Dir, state: &SavedState) -> Result<()> {
     let subdir = sysroot_dir.sub_dir(STATEFILE_DIR)?;
-    let f = {
-        let f = subdir
-            .new_unnamed_file(0o644)
-            .context("creating temp file")?;
-        let mut buff = std::io::BufWriter::new(f);
-        serde_json::to_writer(&mut buff, state)?;
-        buff.flush()?;
-        buff.into_inner()?
-    };
-    let dest_tmp_name = {
-        // expect OK because we just created the filename above from a constant
-        let mut buf = std::ffi::OsString::from(STATEFILE_NAME);
-        buf.push(".tmp");
-        buf
-    };
-    let dest_tmp_name = Path::new(&dest_tmp_name);
-    if subdir.exists(dest_tmp_name)? {
-        subdir
-            .remove_file(dest_tmp_name)
-            .context("Removing temp file")?;
-    }
-    subdir
-        .link_file_at(&f, dest_tmp_name)
-        .context("Linking temp file")?;
-    f.sync_all().context("syncing")?;
-    subdir
-        .local_rename(dest_tmp_name, STATEFILE_NAME)
-        .context("Renaming temp file")?;
+    subdir.write_file_with_sync(STATEFILE_NAME, 0o644, |w| -> Result<()> {
+        serde_json::to_writer(w, state)?;
+        Ok(())
+    })?;
     Ok(())
 }
 
```

File: src/lib.rs [hunk 2]
```diff
@@ -52,6 +54,78 @@ pub trait OpenatDirExt {
...
+        f: F,
+    ) -> Result<T, E>
+    where
+        F: Fn(&mut std::io::BufWriter<std::fs::File>) -> Result<T, E>,
+        E: From<io::Error>,
+    {
+        let mut w = self.new_file_writer(destname, mode)?;
+        match f(&mut w.writer) {
+            Ok(v) => {
+                w.complete_with(|f| {
```
Comment (lucab) [inline src/lib.rs:107]: Who is meant to do the BufWriter flushing in this codepath? I was expecting the library to do it here right before the `complete_with`.

**Comment by lucab** [comment]
One question on buffer flushing in the sync case, PR looks fine otherwise.

File: src/lib.rs (hunk 2 above)
Comment (cgwalters) [reply src/lib.rs:107]: [This bit of code](https://github.com/cgwalters/openat-ext/pull/8/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R313) does it, see [into_inner docs](https://doc.rust-lang.org/std/io/struct.BufWriter.html#method.into_inner).

**Comment by cgwalters** [comment]
OK something not addressed here is automatic handling for mountpoint crossing - I tried that locally but it got much more involved than I thought because in that case the `FileWriter` needs to own the dfd and I need to do some research on how to handle cases where one either borrows or owns data.
